packages/flutter/lib/src/rendering/object.dart

```diff
@@ -651,31 +651,58 @@ class PaintingContext extends ClipContext {
     }
   }
 
-  /// Blend further painting with an alpha value.
+  /// Blend further painting with an opacity value.
   ///
-  /// The `offset` argument indicates an offset to apply to all the children
-  /// (the rendering created by `painter`).
+  /// {@macro flutter.rendering.PaintingContext.pushClipRect.needsCompositing}
+  ///
+  /// The [offset] argument indicates an offset to apply to all the children
+  /// (the rendering created by [painter]).
+  ///
+  /// The [size] is used to compute the painting bounds if a save layer is used.
+  /// If the [size] is not provided, no estimate is used.
   ///
-  /// The `alpha` argument is the alpha value to use when blending the painting
-  /// done by `painter`. An alpha value of 0 means the painting is fully
-  /// transparent and an alpha value of 255 means the painting is fully opaque.
+  /// The [opacity] argument is the opacity value to use when blending the painting
+  /// done by [painter]. An opacity value of 0.0 means the painting is fully
+  /// transparent and an opacity value of 1.0 means the painting is fully opaque.
   ///
-  /// The `painter` callback will be called while the `alpha` is applied. It
+  /// The [painter] callback will be called while the [opacity] is applied. It
   /// is called synchronously during the call to [pushOpacity].
   ///
-  /// {@macro flutter.rendering.PaintingContext.pushClipRect.oldLayer}
+  /// This method will only create a layer if [needsCompositing] is `true` and
+  /// [opacity] is greater than 0 and less than 1.0.
   ///
-  /// A [RenderObject] that uses this function is very likely to require its
-  /// [RenderObject.alwaysNeedsCompositing] property to return true. That informs
-  /// ancestor render objects that this render object will include a composited
-  /// layer, which, for example, causes them to use composited clips.
-  OpacityLayer pushOpacity(Offset offset, int alpha, PaintingContextCallback painter, { OpacityLayer? oldLayer }) {
-    final OpacityLayer layer = oldLayer ?? OpacityLayer();
-    layer
-      ..alpha = alpha
-      ..offset = offset;
-    pushLayer(layer, painter, Offset.zero);
-    return layer;
+  /// {@macro flutter.rendering.PaintingContext.pushClipRect.oldLayer}
+  OpacityLayer? pushOpacity(bool needsCompositing, Offset offset, double opacity, PaintingContextCallback painter, { OpacityLayer? oldLayer, Size? size }) {
```

Comment:
can we move this into the Opacity render object? (we might not be able to, i forget the constraints, but in general the more we can push out of PaintingContext the simpler it gets to eventually remove it entirely)

Reply:
this is used in a few different places, I think the flow layout and a cupertino widget. The code itself isn't that complicated so I could repeat it, or I could move it into a toplevel helper method in proxy_box. WDUT?